core/deployment/src/main/java/io/quarkus/deployment/builditem/StaticBytecodeRecorderBuildItem.java

```diff
@@ -3,15 +3,23 @@
 import io.quarkus.builder.item.MultiBuildItem;
 import io.quarkus.deployment.recording.BytecodeRecorderImpl;
 
-public final class StaticBytecodeRecorderBuildItem extends MultiBuildItem {
+public final class StaticBytecodeRecorderBuildItem extends MultiBuildItem
+        implements Comparable<StaticBytecodeRecorderBuildItem> {
 
     private final BytecodeRecorderImpl bytecodeRecorder;
+    private final int ordinal;
 
-    public StaticBytecodeRecorderBuildItem(BytecodeRecorderImpl bytecodeRecorder) {
+    public StaticBytecodeRecorderBuildItem(BytecodeRecorderImpl bytecodeRecorder, int ordinal) {
```

Comment:
Adding an ordinal here shouldn't be necessary because the steps were already topologically sorted, therefore the build items produced will also be sorted by the same topology.

Reply:
This is an interesting point. You mean it is possible to impose the canonical ordering of the Buildstep executions? I thought this was not possible due to the concurrent nature of the implementation.

Reply:
Well... to refine/correct my thought a bit... step execution has a partial order, with multiple possible correct total orders, hence our original problem, I suppose.  So you are correct that in this case we're trying to pin things down to a single correct total order and to do that requires yet another adjustment.  But sorting build items on an ad-hoc basis isn't a sustainable solution IMO.

We essentially need a way to guarantee that each multi build item type has a total order that reflects the total order of the initial step addition regardless of the order they executed in.  This also has the benefit if not requiring a sort step at the end.  I don't think we need any greater concept of total order beyond this.

I imagine one possible solution would be an array of lists.  The outer array has one "slot" for each step that can contribute the multi type, and the inner lists are the instances which are actually contributed by each step.  Then the array of lists can be flattened into a single "real" list once the producers have completed and the consumers begin.  This is marginally more efficient (in big-O terms) than sorting (`O(n)` vs `O(n log n)`) but also doesn't require items to be comparable.

Reply:
> step execution has a partial order, with multiple possible correct total orders, hence our original problem, I suppose.

Yes, exactly.

> the extension load process is sequental, so for a given build, it should run in the same order each time

I was not aware of this, thanks for pointing out. It is really the case and it makes the whole thing much easier.

> one possible solution would be an array of lists

I tried to implement that idea in aacdcf7 - see the new `Multis` class. It works quite well.